### PR TITLE
Add `Image` validation rule

### PR DESCRIPTION
**Pull request by vjik**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ❌
| Fixed issues  |

### CODE REVIEW

File: src/Rule/ImageHandler.php [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+        }
+
+        if (!$this->isImageFile($filePath)) {
+            return null;
+        }
+
+        /**
+         * @psalm-var (array{0:int,1:int}&array)|null $info Need for PHP 8.0 only
+         */
+        $info = getimagesize($filePath);
```
Comment (xepozz) [inline src/Rule/ImageHandler.php:88]: https://github.com/symfony/symfony/issues/46964

Could you add a comment about HEIC codec?

**Review by xepozz — APPROVED** [review]

File: src/Rule/Image.php [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+ */
+#[Attribute(Attribute::TARGET_PROPERTY | Attribute::IS_REPEATABLE)]
+final class Image implements RuleWithOptionsInterface, SkipOnErrorInterface, WhenInterface, SkipOnEmptyInterface
+{
+    use SkipOnEmptyTrait;
+    use SkipOnErrorTrait;
+    use WhenTrait;
+
+    /**
+     * @param bool|callable|null $skipOnEmpty Whether to skip this rule if the value validated is empty.
```
Comment (samdark) [inline src/Rule/Image.php:31]: Some params are missing from phpdoc.

File: src/Rule/Image.php [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+use Yiisoft\Validator\Rule\Trait\SkipOnEmptyTrait;
+use Yiisoft\Validator\Rule\Trait\SkipOnErrorTrait;
+use Yiisoft\Validator\Rule\Trait\WhenTrait;
+use Yiisoft\Validator\RuleWithOptionsInterface;
+use Yiisoft\Validator\SkipOnEmptyInterface;
+use Yiisoft\Validator\SkipOnErrorInterface;
+use Yiisoft\Validator\WhenInterface;
+
+/**
+ * @see ImageHandler
```
Comment (samdark) [inline src/Rule/Image.php:18]: Need description.

File: src/Rule/ImageHandler.php [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+
+namespace Yiisoft\Validator\Rule;
+
+use Psr\Http\Message\UploadedFileInterface;
+use Yiisoft\Validator\Exception\UnexpectedRuleException;
+use Yiisoft\Validator\Result;
+use Yiisoft\Validator\RuleHandlerInterface;
+use Yiisoft\Validator\ValidationContext;
+
+final class ImageHandler implements RuleHandlerInterface
```
Comment (samdark) [inline src/Rule/ImageHandler.php:13]: Need phpdoc

File: src/Rule/Image.php [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+     */
+    public function __construct(
+        private ?int $width = null,
+        private ?int $height = null,
+        private ?int $minWidth = null,
+        private ?int $minHeight = null,
+        private ?int $maxWidth = null,
+        private ?int $maxHeight = null,
+        private string $notImageMessage = 'The value must be an image.',
+        private string $notExactlyWidthMessage = 'The width of image "{attribute}" must be exactly {exactly, number} {exactly, plural, one{pixel} other{pixels}}.',
```
Comment (arogachev) [inline src/Rule/Image.php:48]: Validating aspect ratio is quite a popular case too. Can be added iteratively in a separate PR.

**Review by arogachev — APPROVED** [review]

File: src/Rule/Image.php (hunk 5 above)
Comment (arogachev) [inline src/Rule/Image.php:48]: Would be also nice to be able to control file size, mime type, etc. But seems like it's better to implement in a separate rule - `File` and use it together.

**Comment by arogachev** [comment]
Docs related additions - #671.

File: CHANGELOG.md [hunk 6]
```diff
@@ -4,6 +4,7 @@
 
 - New #665: Add methods `addErrorWithFormatOnly()` and `addErrorWithoutPostProcessing()` to `Result` object (@vjik)
 - Enh #668: Clarify psalm types in `Result` (@vjik)
+- New #670: Add `Image` validation rule (@vjik)
```
Comment (vjik) [inline CHANGELOG.md:7]:
```suggestion
- New #670: Add `Image` validation rule (@vjik, @arogachev)
```

File: src/Rule/Image/ImageHandler.php [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+        if (empty($imageFilePath)) {
+            $result->addError($rule->getNotImageMessage(), ['attribute' => $context->getTranslatedAttribute()]);
+            return $result;
+        }
+
+        if (!$this->isNeedToValidateDeminisions($rule)) {
+            return $result;
+        }
+
+        $info = $this->imageInfoProvider->get($imageFilePath);
```
Comment (BoShurik) [inline src/Rule/Image/ImageHandler.php:48]: Looks like an over-engineering to me. It's just a `@getimagesize($path)` call

File: src/Rule/Image/ImageHandler.php [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+            $result->addError($rule->getTooLargeHeightMessage(), [
+                'attribute' => $context->getTranslatedAttribute(),
+                'limit' => $rule->getMaxHeight(),
+            ]);
+        }
+
+        return $result;
+    }
+
+    private function isNeedToValidateDeminisions(Image $rule): bool
```
Comment (arogachev) [inline src/Rule/Image/ImageHandler.php:97]:
```suggestion
    private function needValidateDimensions(Image $rule): bool
```

File: src/Rule/Image/Image.php [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+use Yiisoft\Validator\Rule\Trait\SkipOnEmptyTrait;
+use Yiisoft\Validator\Rule\Trait\SkipOnErrorTrait;
+use Yiisoft\Validator\Rule\Trait\WhenTrait;
+use Yiisoft\Validator\RuleWithOptionsInterface;
+use Yiisoft\Validator\SkipOnEmptyInterface;
+use Yiisoft\Validator\SkipOnErrorInterface;
+use Yiisoft\Validator\WhenInterface;
+
+/**
+ * Defines validation options to check that a value is an image with a certain dimensions (optionally).
```
Comment (samdark) [inline src/Rule/Image/Image.php:18]:
```suggestion
 * Defines validation options to check that a value is an image and optionally check its dimensions.
```

File: src/Rule/Image/ImageHandler.php [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+use Psr\Http\Message\UploadedFileInterface;
+use Yiisoft\Validator\Exception\UnexpectedRuleException;
+use Yiisoft\Validator\Result;
+use Yiisoft\Validator\RuleHandlerInterface;
+use Yiisoft\Validator\ValidationContext;
+
+use function is_string;
+
+/**
+ * Validates that a value is an image with a certain dimensions (optionally).
```
Comment (samdark) [inline src/Rule/Image/ImageHandler.php:16]:
```suggestion
 * Validates that a value is an image and optionally checks its dimensions.
```

File: src/Rule/Image/ImageHandler.php (hunk 8 above)
Comment (samdark) [inline src/Rule/Image/ImageHandler.php:97]:
```suggestion
    private function shouldValidateDimensions(Image $rule): bool
```

File: src/Rule/Image.php (hunk 5 above)
Comment (vjik) [reply src/Rule/Image.php]: https://github.com/yiisoft/validator/issues/591

**Pull request merged**